core/peerset/src/lib.rs

```diff
@@ -103,130 +199,149 @@ pub struct PeersetConfig {
+					// let's remove it
+					self.data.discovered.remove_peer(&peer_id);
+
+					// notify that connection has been made
+					self.message_queue.push_back(Message::Connect(peer_id));
+					return;
+				},
+				Err(SlotError::AlreadyConnected(_)) => {
+					return;
+				}
+				Err(SlotError::MaxConnections(_)) => {
+					self.data.discovered.add_peer(peer_id, SlotType::Reserved);
+					return;
+				}
+				Err(SlotError::DemandReroute { disconnect, ..}) => {
+					// disconnect not reserved node
+					self.data.out_slots.clear_slot(&disconnect);
+					self.message_queue.push_back(Message::Drop(disconnect));
+					// on the next loop iteration we should connect to the peer
+				}
 			}
 		}
 	}
 
-	/// Remove a previously-added reserved peer.
-	///
-	/// Has no effect if the node was not a reserved peer.
-	pub fn remove_reserved_peer(&self, peer_id: &PeerId) {
-		let mut inner = self.inner.lock();
-		inner.reserved.remove(peer_id);
+	fn on_remove_reserved_peer(&mut self, peer_id: PeerId) {
+		self.data.in_slots.mark_not_reserved(&peer_id);
+		self.data.out_slots.mark_not_reserved(&peer_id);
+		self.data.discovered.mark_not_reserved(&peer_id);
+		if self.data.reserved_only {
+			if self.data.in_slots.contains(&peer_id) || self.data.out_slots.contains(&peer_id) {
+				self.data.in_slots.clear_slot(&peer_id);
+				self.data.out_slots.clear_slot(&peer_id);
+				self.message_queue.push_back(Message::Drop(peer_id))
+			}
+		}
 	}
 
-	/// Sets whether or not the peerset only has connections .
-	pub fn set_reserved_only(&self, reserved_only: bool) {
-		let mut inner = self.inner.lock();
-		let inner = &mut *inner;	// Fixes a borrowing issue.
-		inner.reserved_only = reserved_only;
-
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
 		// Disconnect non-reserved nodes.
-		if reserved_only {
-			for slot in inner.out_slots.iter_mut().chain(inner.in_slots.iter_mut()) {
-				if let Some(peer) = slot.as_ref() {
-					if inner.reserved.contains(peer) {
-						continue;
-					}
-
-					let _ = self.tx.unbounded_send(Message::Drop(peer.clone()));
-				}
-
-				*slot = None;
+		self.data.reserved_only = reserved_only;
+		if self.data.reserved_only {
+			for peer_id in self.data.in_slots.clear_common_slots().into_iter().chain(self.data.out_slots.clear_common_slots().into_iter()) {
+				// peer will be removed from `in_slots` or `out_slots` in `on_dropped` method
+				self.data.in_slots.clear_slot(&peer_id);
+				self.message_queue.push_back(Message::Drop(peer_id.clone()));
+				self.data.discovered.add_peer(peer_id, SlotType::Common);
 			}
+		} else {
+			self.alloc_slots();
 		}
 	}
 
-	/// Reports an adjustement to the reputation of the given peer.
-	pub fn report_peer(&self, _peer_id: &PeerId, _score_diff: i32) {
-		// This is not implemented in this dummy implementation.
-	}
-}
+	fn on_report_peer(&mut self, peer_id: PeerId, score_diff: i32) {
+		let score = self.data.scores.entry(peer_id.clone()).or_default();
```

Comment:
I think `scores` grows forever, right? For long running nodes it will take up a lot of memory.

Reply:
true, we can use some `lru_cache` here.

The same problem exists with `discovered`. Should we limit the number of peers over there as well?